internal/profiling/profiling.go

```diff
@@ -0,0 +1,213 @@
+// to mark the end. The return value is not a pointer to a Timer object because
+// the internal slice may be re-allocated freely, which would make the
+// reference to pointers of the past obsolete.
+//
+// Why a slice of timer? To make NewTimer allocation-free, we don't allocate a
+// new timer every time. Instead a slice of timers large enough to hold most
+// data is created and elements from this slice are used. This should put less
+// pressure on the GC too.
+func (stat *Stat) NewTimer(timerTag string) (index int) {
+	if stat != nil {
+		// TODO(adtac): Make this lock-free? We don't expect much contention on
+		// this, so I don't predict a significant improvement in performance. Using
+		// a lock-free, concurrent queue will require us to write our own
+		// reallocation and copy, which is probably unnecessary code complexity if
+		// the improvement isn't significant.
+		stat.mu.Lock()
+		stat.Timers = append(stat.Timers, Timer{
+			TimerTag: timerTag,
+			Begin:    time.Now(),
+			GoId:     goId(),
+		})
+		index = len(stat.Timers) - 1
+		stat.mu.Unlock()
+	}
+
+	return
+}
+
+// Egress marks the completion of a given timer within a stat.
+func (stat *Stat) Egress(index int) {
+	if stat != nil && index < len(stat.Timers) {
+		stat.Timers[index].End = time.Now()
+	}
+}
+
+// AppendTimer appends a given Timer object to the internal slice of timers. A
+// deep copy of the timer is made (i.e. no reference is retained to this
+// pointer) and the user is expected to lose their reference to the timer to
+// allow the Timer object to be garbage collected.
+func (stat *Stat) AppendTimer(timer *Timer) (index int) {
+	if stat != nil && timer != nil {
+		stat.mu.Lock()
+		stat.Timers = append(stat.Timers, *timer)
+		index = len(stat.Timers) - 1
+		stat.mu.Unlock()
+	}
+
+	return
+}
+
+// ConnectionCounters count the number of connections. This counter is embedded
+// within a StreamStat's Metadata along with each stream's stream ID to
+// uniquely identify a query.
+var ServerConnectionCounter uint64
+var ClientConnectionCounter uint64
+
+// Stats for the last defaultStreamStatsBufsize RPCs will be stored in memory.
+// This is can be configured by the registering server at profiling service
+// initialisation with google.golang.org/grpc/profiling/service.ProfilingConfig
+const defaultStreamStatsSize uint32 = 16 << 10
+
+var statsInitialised int32
+
+// StreamStats is a CircularBuffer containing data from the last N RPC calls
+// served, where N is set by the user. This will contain both server stats and
+// client stats (but each stat will be tagged with whether it's a server or a
+// client in its StatTag).
+var StreamStats *circularBuffer
+
+// InitStats initialises all the relevant Stat objects. Must be called exactly
+// once per lifetime of a process; calls after the first one are ignored.
+func InitStats(streamStatsSize uint32) (err error) {
+	if !atomic.CompareAndSwapInt32(&statsInitialised, 0, 1) {
```

Comment:
How about using a sync.One to perform initialization that you want to be done only one?

Reply:
I think a CAS is much cleaner and simpler than defining a func and calling sync.Once. I'll admit I didn't know about sync.Once before, but that looks more useful when there are multiple concurrent goroutines contending for the same thing.

Looking at [the source](https://golang.org/src/sync/once.go#L40), they even talk about not using a CAS because they have different goals -- we explicitly want the function to do nothing and return when things are already initialised.